lib/std/sort.zig

```diff
@@ -502,6 +509,267 @@ test "binarySearch" {
     );
 }
 
+/// Returns the index of the first element in `items` greater than or equal to `key`,
+/// or `items.len` if all elements are less than `key`.
+///
+/// `items` must be sorted in ascending order with respect to `compareFn`.
+///
+/// O(log n) complexity.
+pub fn lowerBound(
+    comptime T: type,
+    key: anytype,
+    items: []const T,
+    context: anytype,
+    comptime compareFn: fn (context: @TypeOf(context), lhs: @TypeOf(key), rhs: T) math.Order,
+) usize {
+    var left: usize = 0;
+    var right: usize = items.len;
+
+    while (left < right) {
+        const mid = left + (right - left) / 2;
+        if (compareFn(context, items[mid], key) == .lt) {
```

Comment:
Didn't this used to be a `lessThan` function before? By changing it to `compareFn`, now it does potentially 2 comparisons instead of 1 in each iteration.

It looks like all the functions could have simply taken a `lessThan` function, no?

Reply:
They did, but I changed them because your comment: https://github.com/ziglang/zig/pull/18718#discussion_r1468955495, I must have misunderstood! So they should all take `lessThan`? I realize I may have confused you because in my previous commit `equalRange` took two functions (`lessThan` and `greaterThan`) but that is not necessary and all three functions can work with the same `lessThan` function.

Reply:
Yes I think you made other improvements along the way that now reveal it can be the ideal case of only taking one `lessThan` function. Sorry for the confusion. It was not obvious to me that it would work this way at first.

1. Most preferred scenario: only `lessThan`
2. Acceptable: only `compare`
3. No good: both `lessThan` and `greaterThan`